sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -217,68 +217,54 @@ export class EventHubSender extends LinkEntity {
       abortSignal?: AbortSignalLike;
     } = {}
   ): Promise<number> {
-    const abortSignal = options.abortSignal;
-    const retryOptions = options.retryOptions || {};
     if (this.isOpen()) {
       return this._sender!.maxMessageSize;
     }
-    return new Promise<number>(async (resolve, reject) => {
-      const rejectOnAbort = (): void => {
-        const desc: string = `[${this._context.connectionId}] The create batch operation has been cancelled by the user.`;
-        // Cancellation is user-intented, so treat as info instead of warning.
-        logger.info(desc);
-        const error = new AbortError(`The create batch operation has been cancelled by the user.`);
-        reject(error);
-      };
-
-      const onAbort = (): void => {
-        if (abortSignal) {
-          abortSignal.removeEventListener("abort", onAbort);
-        }
-        rejectOnAbort();
-      };
-
-      if (abortSignal) {
-        // the aborter may have been triggered between request attempts
-        // so check if it was triggered and reject if needed.
-        if (abortSignal.aborted) {
-          return rejectOnAbort();
-        }
-        abortSignal.addEventListener("abort", onAbort);
-      }
+
+    const retryOptions = options.retryOptions || {};
+    const timeoutInMs = getRetryAttemptTimeoutInMs(retryOptions);
+    retryOptions.timeoutInMs = timeoutInMs;
+    const senderOptions = this._createSenderOptions(timeoutInMs);
+
+    const createLinkPromise = async (): Promise<void> => {
       try {
-        logger.verbose(
-          "Acquiring lock %s for initializing the session, sender and " +
-            "possibly the connection.",
-          this.senderLock
-        );
-        const senderOptions = this._createSenderOptions(Constants.defaultOperationTimeoutInMs);
-        await defaultLock.acquire(this.senderLock, () => {
-          const config: RetryConfig<void> = {
-            operation: () => this._init(senderOptions),
-            connectionId: this._context.connectionId,
-            operationType: RetryOperationType.senderLink,
-            abortSignal: abortSignal,
-            retryOptions: retryOptions
-          };
-
-          return retry<void>(config);
+        await waitForTimeoutOrAbortOrResolve({
+          actionFn: () => {
+            return defaultLock.acquire(this.senderLock, () => {
+              return this._init(senderOptions);
+            });
+          },
+          abortSignal: options?.abortSignal,
+          timeoutMs: timeoutInMs,
+          timeoutMessage:
+            `[${this._context.connectionId}] Sender "${this.name}" ` +
+            `with address "${this.address}", cannot be created right now, due ` +
+            `to operation timeout.`
         });
-        resolve(this._sender!.maxMessageSize);
       } catch (err) {
```

Comment:
I was wondering... is it worth adding a catch block here? The `retry` method anyway catches any error, translates it and then determines if it should be retried. So, arent we well off skipping this try/catch and adding one around the retry method instead if we wanted specific logging?

Reply:
> I think that's what we were doing prior to this change as well

Yes. Additionally, all the retries were inside the lock. Am guessing we don't need to hold on to the lock in between retries?

Reply:
Updated in https://github.com/Azure/azure-sdk-for-js/pull/13826/commits/ec7f720c474b9973d4f14ebe6388b5fda98dd680

Reply:
> Yes. Additionally, all the retries were inside the lock. Am guessing we don't need to hold on to the lock in between retries?

Yeah, I think now that the timeout applies to the time it takes to acquire the lock as well, giving other operations the chance to acquire the lock after a failed, retryable operation is the 'good citizen' thing to do.